File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.py
import sys
input = sys.stdin.read
sys.setrecursionlimit(10**7)

# 노드 생성
class Node(object):   
  def __init__(self, data):       
    self.data = data        
    self.left = None
    self.right = None

# 이진 트리 생성
class BinarySearchTree(object):    
  def __init__(self):        
    self.root = None
  
  # 노드 삽입 (반복적 방식)
  def insert(self, data):
    new_node = Node(data)
    if self.root is None:
      self.root = new_node
      return
    current = self.root
    while True:
      if data < current.data:
        if current.left is None:
          current.left = new_node
          break
        current = current.left
      else:
        if current.right is None:
          current.right = new_node
          break
        current = current.right

# 후위 순회(Postorder)
def postorder(node):
  if node is not None:
    postorder(node.left)
    postorder(node.right)
    print(node.data)

# 입력 처리 및 트리 생성
data = input().strip().split()
tree = BinarySearchTree()

# 첫 번째 값을 루트로 삽입
tree.insert(int(data[0]))

# 나머지 값들을 트리에 삽입
for num in data[1:]:
  tree.insert(int(num))

postorder(tree.root)

File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.java
import java.io.*;

class Node {
    int key;
    Node left;
    Node right;

    public Node(int key) {
        this.key = key;
        this.left = null;
        this.right = null;
    }
}

class BinarySearchTree {
    Node root;

    public BinarySearchTree() {
        this.root = null;
    }

    public void insert(int key) {
        root = insertRec(root, key);
    }

    private Node insertRec(Node root, int key) {
        if (root == null) {
            root = new Node(key);
            return root;
        }

        if (key < root.key) root.left = insertRec(root.left, key);
        else if (key > root.key) root.right = insertRec(root.right, key);

        return root;
    }

    public void postOrder(Node node) {
        if (node == null) return;

        postOrder(node.left);
        postOrder(node.right);
        System.out.println(node.key);
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String line;
        BinarySearchTree bst = new BinarySearchTree();

        while ((line = br.readLine()) != null && !line.isEmpty()) {
            int num = Integer.parseInt(line);
            bst.insert(num);
        }

        bst.postOrder(bst.root);
    }
}